pkg/k8s/client/object_tracker.go

```diff
@@ -0,0 +1,617 @@
+			o:        obj})
+	}
+	return nil
+}
+
+// Apply implements testing.ObjectTracker.
+func (s *statedbObjectTracker) Apply(gvr schema.GroupVersionResource, applyConfiguration runtime.Object, ns string, opts ...metav1.PatchOptions) error {
+	log := s.log.With(
+		logfieldClientset, s.domain,
+		logfields.Object, applyConfiguration)
+
+	applyConfigurationMeta, err := meta.Accessor(applyConfiguration)
+	if err != nil {
+		log.Debug("Apply", logfields.Error, err)
+		return err
+	}
+
+	obj, err := s.Get(gvr, ns, applyConfigurationMeta.GetName(), metav1.GetOptions{})
+	if err != nil {
+		log.Debug("Apply", logfields.Error, err)
+		return err
+	}
+
+	old, err := json.Marshal(obj)
+	if err != nil {
+		log.Debug("Apply", logfields.Error, err)
+		return err
+	}
+
+	// reset the object in preparation to unmarshal, since unmarshal does not guarantee that fields
+	// in obj that are removed by patch are cleared
+	value := reflect.ValueOf(obj)
+	value.Elem().Set(reflect.New(value.Type().Elem()).Elem())
+
+	// For backward compatibility with behavior 1.30 and earlier, continue to handle apply
+	// via strategic merge patch (clients may use fake.NewClientset and ManagedFieldObjectTracker
+	// for full field manager support).
+	patch, err := json.Marshal(applyConfiguration)
+	if err != nil {
+		log.Debug("Apply", logfields.Error, err)
+		return err
+	}
+	mergedByte, err := strategicpatch.StrategicMergePatch(old, patch, obj)
+	if err != nil {
+		log.Debug("Apply", logfields.Error, err)
+		return err
+	}
+	if err = json.Unmarshal(mergedByte, obj); err != nil {
+		log.Debug("Apply", logfields.Error, err)
+		return err
+	}
+
+	err = s.Update(gvr, obj, ns)
+	s.log.Debug("Apply", logfields.Error, err)
+	return err
+}
+
+// Create implements testing.ObjectTracker.
+func (s *statedbObjectTracker) Create(gvr schema.GroupVersionResource, obj runtime.Object, ns string, opts ...metav1.CreateOptions) error {
+	log := s.log.With(
+		logfieldClientset, s.domain,
+		logfields.Object, obj)
+
+	newMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+	if len(newMeta.GetNamespace()) == 0 {
+		newMeta.SetNamespace(ns)
+	}
+	wtxn := s.db.WriteTxn(s.tbl)
+	_, found, _ := s.tbl.Insert(wtxn, object{
+		objectId: newObjectId(s.domain, gvr, ns, newMeta.GetName()),
+		o:        obj})
+	if found {
+		wtxn.Abort()
+		gr := gvr.GroupResource()
+		err := apierrors.NewAlreadyExists(gr, newMeta.GetName())
+		log.Debug("Create", logfields.Error, err)
+	}
+	log.Debug("Create")
+	wtxn.Commit()
+	return nil
+}
+
+// Delete implements testing.ObjectTracker.
+func (s *statedbObjectTracker) Delete(gvr schema.GroupVersionResource, ns string, name string, opts ...metav1.DeleteOptions) error {
+	log := s.log.With(
+		logfieldClientset, s.domain,
+		logfields.K8sNamespace, ns,
+		logfields.Name, name)
+
+	wtxn := s.db.WriteTxn(s.tbl)
+	_, found, _ := s.tbl.Delete(wtxn, object{objectId: newObjectId(s.domain, gvr, ns, name)})
```

Comment:
Surprised that this isn't idempotent deletion - but I guess this comes from upstream?

Reply:
Yep also upstream. I'm keeping all the same errors as there.